File: default/task_queue_client.py
import logging
from google.appengine.api import urlfetch
import json
from os import path
import urllib

class Queue():
  def __init__(self, api_url):
    self.api_url = api_url
  
  def list(self):
    """ returns the list of tasks (filters will be added later) """
    response = urlfetch.fetch(
        url=path.join(self.api_url, 'list'),
        method=urlfetch.GET
    )
    if response.status_code != 200:
      handleError(response)
    return json.loads(response.content)

  
  def lease(self, lease_duration=None, tags=None):
    """ Lease a task from the queue
    
    Keyword Arguments:
      lease_duration {int} -- duration in seconds before the task is re-added to
          the queue (if not marked as 'DONE'). Set to None for the API default
          value
      tags {[string]} -- a list of tags to filter the tasks. Set to None for the
          API default value

    Returns:
      {dict|None} -- The leased task object, unless there are none.
    """
    params = {}
    if tags: params['tags'] = ', '.join(tags)
    if lease_duration: params['lease_duration'] = lease_duration
    param_string = urllib.urlencode(params)
    
    request_url = path.join(self.api_url, 'lease?'+param_string)

    response = urlfetch.fetch(url=request_url, method=urlfetch.GET)
    if response.status_code == 204:
      return None
    if response.status_code != 200:
      handleError(response)

    return json.loads(response.content)

  def getTask(self, task_id):
    """Returns the task object for the given ID
    
    Arguments:
      task_id {string} -- a task id
    
    Raises:
      Exception -- raised when the task id is invalid, or does not exist
    
    Returns:
      {dict} -- the task object for the given ID
    """
    response = urlfetch.fetch(
        url=path.join(self.api_url, 'task'),
        method=urlfetch.GET
    )
    if response.status_code != 200:
      handleError(response)
    return json.loads(response.content)

  def appendTask(self, payload=None, tags=None, max_attempts=None):
    """Creates a task and appends it to the queue
    
    Keyword Arguments:
      payload {dict} -- json-serializable payload for the task
      tags {[string]} -- a list of tags to itentify groups of tasks within the queue
      max_attempts {int} -- how many attempts should the task be tried before being declared 'FAILED'
    
    Returns:
      {dict} -- The created Task object's key
    """

    request_data = {}
    if payload!=None: request_data["payload"] = payload
    if tags!=None: request_data["tags"] = tags
    if max_attempts!=None: request_data["max_attempts"] = max_attempts
    
    response = urlfetch.fetch(
        url=path.join(self.api_url, 'task'),
        method=urlfetch.POST,
        payload=json.dumps(request_data),
    )
    if response.status_code != 200:
      handleError(response)
    return json.loads(response.content)["task_key"]
  
  def updateTask(self, task_id, status=None, response=None):
    """ Updates a task's properties
    
    Arguments:
      task_id {string} -- the id for the task to be modified
    
    Keyword Arguments:
      status {int} -- The status of the task
      response {dict} -- The 'response' object for the task.
    
    Returns:
      {dict} -- The modified task object
    """

    request_data = {}
    if status!=None: request_data["status"] = status
    if response!=None: request_data["response"] = response

    response = urlfetch.fetch(
        url=path.join(self.api_url, 'task', task_id),
        method=urlfetch.PUT,
        payload=json.dumps(request_data)
    )
    if response.status_code != 200:
      handleError(response)
    return json.loads(response.content)

def handleError(response):
  try:
    raise Exception(json.loads(response.content))
  except:
    raise Exception(response.content)

File: render_service/task_queue_client.js
const fetch = require('node-fetch');
const URL = require('url').URL;
var querystring = require("querystring");

/**
 * An enum representing a task's status
 * @typedef {('PENDING'|'RUNNING'|'DONE'|'FAILED')} Status
 */

/**
 * An object representing a task
 * @typedef {Object} Task
 * @property {Status} status - the current status of the task
 * @property {Number} create_time - the timestamp for when the task was created
 * @property {Object} payload - the payload for the task
 * @property {Object} response - the response for the task (should be empty on lease)
 * @property {String[]} tags - a list of arbitrary string tags
 * @property {String} api_version - the api version: major.minor.patch
 * @property {String} key - the key (=id) of the task
 * @property {String} attempt_count - number of times this task has already been attempted
 * @property {String} max_attempts - max number of attempts for the task before declaring it 'FAILED'
 * @property {Number} lease_timeout - The timestamp when the current lease will expire
 * @property {String} lease_timeout_str - A human-readable version of the lease_timeout property
 * 
 */

class TaskQueue{
  constructor(api_url) {
    this.api_url = api_url;
  }

  /**
   * returns the list of tasks (filters will be added later)
   * @return {Task[]} the list of tasks */
  async list() {
    let url = new URL(this.api_url)
    url.pathname = "/list";

    let response = await fetch(url, {method: 'GET'});
    let data = await response.json();
    
    if (response.status != 200) {
      throw new Error(JSON.stringify(data));
    } else {
      return data;
    }
  }

  /**
   * 
   * @param {Number} lease_duration - duration in seconds before the task is
   *      re-added to the queue (if not marked as 'DONE'). Set to None for the
   *      API default value
   * @param {String[]} tags - A list of tags to filter the tasks. Set to None
   *      for the API default value
   * @returns {Task} The leased task, or null if none is available
   */
  async lease(lease_duration=null, tags=null){
    let url = new URL(this.api_url)
    url.pathname = "/lease";
    if(tags) url.searchParams.set('tags', tags.join(','));
    if(lease_duration) url.searchParams.set('lease_duration', lease_duration);

    let response = await fetch(url, {method:'GET'});

    if (response.status == 204) {
      return null;
    } else if (response.status == 200) {
      return await response.json();
    } else {
      throw new Error(await response.text());
    }
  }

  /**
   * Get a specific task
   * @param {String} task_id - the ID of the task
   * @returns {Task} the task information
   */
  async getTask(task_id) {
    let url = new URL(this.api_url)
    url.pathname = "/task/"+task_id;

    let response = await fetch(url, {method:'GET'});

    if (response.status == 200) {
      return await response.json();
    } else {
      throw new Error(await response.text());
    }
  }

  /**
   * Creates a task and appends it to the queue
   * @param {Object} payload - the payload for the task
   * @param {String[]} tags - a list of arbitrary string tags
   * @param {Number} max_attempts - max number of attempts for the task before declaring it 'FAILED'
   * @returns {String} - the created Task's key
   */
  async appendTask(payload=null, tags=null, max_attempts=null) {
    let url = new URL(this.api_url)
    url.pathname = "/task";

    let request_data = {};
    if(payload!=null) request_data["payload"] = payload;
    if(tags!=null) request_data["tags"] = tags;
    if(max_attempts!=null) request_data["max_attempts"] = max_attempts;

    let response = await fetch(url, {
      method:'POST',
      body: JSON.stringify(request_data)
    });

    if (response.status == 200) {
      let data = await response.json();
      return data.task_key;
    } else {
      throw new Error(await response.text());
    }
  }

  /**
   * Updates a task. Note that only some of the task's properties can be changed
   *  - status
   *  - response
   * @param {Task} task - The task to be updated, with its modified properties
   */
  async updateTask(task) {
    let url = new URL(this.api_url)
    url.pathname = "/task/"+task.key;

    let request_data = {};
    request_data.status = task.status;
    request_data.response = task.response;

    let response = await fetch(url, {
      method:'PUT',
      body: JSON.stringify(request_data)
    });

    if (response.status == 200) {
      return await response.json();
    } else {
      throw new Error(await response.text());
    }
  }
}

module.exports = TaskQueue